### PR TITLE
Adding role reference field on the OG membership entity.

### PR DESCRIPTION
**Pull request by RoySegall**
#156

I start working on the role field on the OG membership entity. There are couple of problems:
- [ ] When Drupal create for me the external table of the role referencing the property `roles_target_id` is an integer while we need it as a string since the ID of the config entity goes like this: `entity_test-group-1-content_editor`. I need to find a way to alter the schema structure.
- [ ] When using the `$this->get('roles')->value` or `$this->get('roles')->entity` the field does not give me all the loaded entities object as it should be.

### CODE REVIEW

**Comment by RoySegall** [comment]
OK. I handled both of the issues by creating a new entity reference item.

**Comment by RoySegall** [comment]
And the test are passing without doing any thing. Wicked :)

File: src/Entity/OgMembership.php [hunk 1]
```diff
@@ -196,6 +245,14 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
       ->setDescription(t("The state of the group content."))
       ->setDefaultValue(TRUE);
 
+    $fields['roles'] = BaseFieldDefinition::create('og_role_reference')
+      ->setLabel(t('Roles'))
+      ->setDescription(t("The user's group's roles."))
+      ->setCardinality(FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED)
+      ->setDefaultValue(OgGroupAudienceHelper::DEFAULT_FIELD)
```
Comment (amitaibu) [inline src/Entity/OgMembership.php]: This looks weird,  why do we need it?

File: src/Entity/OgMembership.php [hunk 2]
```diff
@@ -157,6 +159,53 @@ public function getType() {
...
 
   /**
+   * Set the group's role's for the current user group membership's.
+   *
+   * @param $roles
+   *   List of og roles ids.
+   *
+   * @return OgMembership
+   */
+  public function setRoles($roles) {
```
Comment (amitaibu) [inline src/Entity/OgMembership.php]: Maybe better role_ids?

**Comment by RoySegall** [comment]
@amitaibu @damiankloip @chx @pfrenssen Can you have a second look on this one? 

10x.

File: src/Entity/OgMembership.php [hunk 3]
```diff
@@ -157,6 +159,95 @@ public function getType() {
...
+   * @param $return_ids
+   *   Determine if we need to return the role IDs or the roles object's.
+   *
+   * @return OgRole[]
+   */
+  public function getRoles($return_ids = FALSE) {
+    $roles = $this->get('roles')->referencedEntities();
+
+    if ($return_ids) {
+      $rids = [];
```
Comment (chx) [inline src/Entity/OgMembership.php]: `return array_map(function ($role) { return $role->id() ; }, $roles);`

File: src/Entity/OgMembership.php [hunk 4]
```diff
@@ -157,6 +159,95 @@ public function getType() {
...
+   * Checks if the user has a permission inside the group.
+   *
+   * @param $permission
+   *   The name of the permission.
+   *
+   * @return bool
+   */
+  public function hasPermission($permission) {
+    $roles = $this->getRoles();
+
```
Comment (chx) [inline src/Entity/OgMembership.php]: return array_filter($this->getRoles(), function ($role) use ($permission) {return $role->hasPermission($permission);} )

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    parent::setUp();
+
+    // Installing needed schema.
+    $this->installConfig(['og']);
+    $this->installEntitySchema('og_membership');
+    $this->installEntitySchema('user');
+    $this->installEntitySchema('node');
+    $this->installSchema('system', 'sequences');
+
+    $this->groupBundle = Unicode::strtolower($this->randomMachineName());
```
Comment (chx) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php]: why strtolower'ing it? is something depending on case?

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      'uid' => $this->user->id(),
+      'type' => $this->groupBundle,
+    ]);
+  }
+
+  /**
+   * Testing OG membership role referencing.
+   */
+  public function testRoleCreate() {
+    // Creating a content editor role.
```
Comment (chx) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php]: OMG this test is beautiful.

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php (hunk 5 above)
Comment (RoySegall) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php]: Ummm... Just took it from another tests. Looks legit.

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php (hunk 6 above)
Comment (RoySegall) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php:75]: 10x :)

Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php:75]: wow @RoySegall , getting an "OMG" from @chx is worth _lots_ of karma points! :)

**Comment by chx** [comment]
We do not yet have this tied into the entity checking system, do we? Or did I miss something?

**Comment by RoySegall** [comment]
Indeed. This is actually the next step. After this one will merged i'll handle the entity checking system.

**Comment by RoySegall** [comment]
OK. This is ready for another look.

**Comment by amitaibu** [comment]
Test seem to be failing

**Comment by RoySegall** [comment]
I re-triggered the failing suite. Now it's OK.

File: src/Entity/OgMembership.php [hunk 7]
```diff
@@ -157,6 +159,91 @@ public function getType() {
...
+   */
+  public function addRole($role_id) {
+    $rids = $this->getRolesIds();
+    $rids[] = $role_id;
+
+    return $this->setRoles(array_unique($rids));
+  }
+
+  /**
+   * Revoking a role from the user's membership's.
```
Comment (amitaibu) [inline src/Entity/OgMembership.php]: user's membership's => from the OG membership

File: src/Entity/OgMembership.php [hunk 8]
```diff
@@ -196,6 +283,12 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
       ->setDescription(t("The state of the group content."))
       ->setDefaultValue(TRUE);
 
+    $fields['roles'] = BaseFieldDefinition::create('entity_reference')
+      ->setLabel(t('Roles'))
+      ->setDescription(t("The user's group's roles."))
```
Comment (amitaibu) [inline src/Entity/OgMembership.php]: Maybe better: `The OG roles related to an OG membership entity`.

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+      ->setLabel('Group member');
+    $group_member->save();
+
+    /** @var OgMembership $membership */
+    $membership = OgMembership::create(['type' => OgMembershipInterface::TYPE_DEFAULT]);
+    $membership
+      ->setEntityType('node')
+      ->setEntityId($this->group->id())
+      ->setUser($this->user)
+      ->setRoles([$content_editor->id()])
```
Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php:101]: `// Assign only the content editor role for now.`

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      ->save();
+
+    $roles_ids = $membership->getRolesIds();
+    $this->assertTrue(in_array($content_editor->id(), $roles_ids), 'The membership own the content editor role.');
```
Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php]: The membership own => The membership has

Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php]: (here and below)

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    $roles_ids = $membership->getRolesIds();
+
+    $this->assertTrue(in_array($content_editor->id(), $roles_ids), 'The membership own the content editor role.');
+    $this->assertTrue(in_array($group_member->id(), $roles_ids), 'The membership own the group member role.');
+
+    // Remove a role.
+    $membership->revokeRole($content_editor->id());
+
+    $roles_ids = $membership->getRolesIds();
+    $this->assertFalse(in_array($content_editor->id(), $roles_ids), 'The membership does not own content editor role after it revoked.');
```
Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php]: does not own => does not have the content editor role after is has been revoked.

**Comment by amitaibu** [comment]
Looks good. Added some nitpick comments.

File: src/Entity/OgMembership.php [hunk 12]
```diff
@@ -157,6 +159,91 @@ public function getType() {
...
+    return $this->get('roles')->referencedEntities();
+  }
+
+  /**
+   * Get list of OG role IDs.
+   *
+   * @return array
+   *   List of OG roles ids.
+   */
+  public function getRolesIds() {
```
Comment (chx) [inline src/Entity/OgMembership.php]: I'd call this `getRoleIds`.

File: src/Entity/OgMembership.php [hunk 13]
```diff
@@ -157,6 +159,91 @@ public function getType() {
...
 
   /**
+   * Set the group's role's for the current user group membership's.
+   *
+   * @param $roles_ids
+   *   List of og roles ids.
+   *
+   * @return OgMembership
+   */
+  public function setRoles($roles_ids) {
```
Comment (chx) [inline src/Entity/OgMembership.php]: `$role_ids`

**Comment by RoySegall** [comment]
@amitaibu Applied the fixes. This is ready to go.

File: src/Entity/OgMembership.php (hunk 12 above)
Comment (amitaibu) [inline src/Entity/OgMembership.php:221]: @RoySegall ^^

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php (hunk 9 above)
Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php:101]: @RoySegall ^^

Comment (RoySegall) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php:101]: adding_role_reference_field_on_the_og_membership_entity__by_roysegall_ _pull_request__157_ _amitaibu_og

File: src/Entity/OgMembership.php (hunk 12 above)
Comment (RoySegall) [inline src/Entity/OgMembership.php:221]: adding_role_reference_field_on_the_og_membership_entity__by_roysegall_ _pull_request__157_ _amitaibu_og

File: tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php (hunk 9 above)
Comment (amitaibu) [inline tests/src/Kernel/Entity/OgMembershipRoleReferenceTest.php:101]: can you move it above `->setRoles([$content_editor->id()])`

**Comment by RoySegall** [comment]
@amitaibu I relocated the comment.

**Pull request merged**